core/src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -621,14 +622,18 @@ public void close() {
     public static abstract class Operation {
         private final Term uid;
         private long version;
+        private long seqNo;
         private final VersionType versionType;
         private final Origin origin;
         private Translog.Location location;
         private final long startTime;
         private long endTime;
 
-        public Operation(Term uid, long version, VersionType versionType, Origin origin, long startTime) {
+        public Operation(Term uid, long seqNo, long version, VersionType versionType, Origin origin, long startTime) {
             this.uid = uid;
+            assert origin != Origin.PRIMARY || seqNo == -1l : "seqNo should not be set when origin is PRIMARY";
+            assert origin == Origin.PRIMARY || seqNo > -1l : "seqNo should  be set when origin is not PRIMARY";
```

Comment:
I'd consider replacing the usage of `-1l` in this line and the prior with the field [proposed previously](https://github.com/elastic/elasticsearch/pull/14651/files#r45225330).